### PR TITLE
New blockchain

### PR DESCRIPTION
**Pull request by Ronserruya**
# Client work is not finished and tested, do not merge

#### Main purpose:
Add routing between two payment services (one on the old blockchain and one on the new), additionally, create new endpoints:

* Find out the blockchain version for an app
* Whitelist a transaction
 
#### Technical description:
* New payment service URL updated in the config
* The marketplace nows checks what blockchain does an app work with, and routes the request to the appropriate payment service.
* When returning orders to the client, the expected blockchain version is also including.
* Added an endpoint to whitelist transactions, which sends the decoded transaction and the expected contents to the payment service for whitelisting.

### CODE REVIEW

File: scripts/src/public/routes/orders.ts [hunk 1]
```diff
@@ -115,3 +119,23 @@ export const getOrderHistory = async function(req: GetOrderHistoryRequest, res:
 	const orderList = await getOrderHistoryService(req.context.user!.id, filters, req.logger);
 	res.status(200).send(orderList);
 } as any as RequestHandler;
+
+export type WhitelistTransactionRequest = Request & {
+	params: {
+		order_id: string;
+	},
+	tx_envelope: string;
```
Comment (nitzantomer) [inline scripts/src/public/routes/orders.ts]: where are these two coming from? how are they added to the request object?

File: scripts/src/public/services/orders.ts [hunk 2]
```diff
@@ -78,6 +78,15 @@ export async function getOrder(orderId: string, logger: LoggerInstance): Promise
 	return orderDbToApi(order);
 }
 
+export async function getOpenOrder(orderId: string): Promise<db.MarketplaceOrder| db.ExternalOrder> {
+	const order = await db.Order.getOne(orderId, "opened") as db.MarketplaceOrder | db.ExternalOrder;
+
+	if (!order) {
```
Comment (nitzantomer) [inline scripts/src/public/services/orders.ts]: why?
won't it be better to return `Promise<db.MarketplaceOrder| db.ExternalOrder | undefined>` and then whoever called this function will deal with it?

File: scripts/src/models/offers.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ export type BlockchainData = {
 	transaction_id?: string;
 	sender_address?: string;
 	recipient_address?: string;
+	blockchain_version?: string;
```
Comment (yosriz) [inline scripts/src/models/offers.ts]: have it as an enum (`export type XXX = "3" | "2"`

File: scripts/src/public/routes/orders.ts (hunk 1 above)
Comment (Ronserruya) [reply scripts/src/public/routes/orders.ts]: They get sent by the client as the json data in the post request. Though I didnt test with a client yet so im not sure it works

Comment (nitzantomer) [reply scripts/src/public/routes/orders.ts]: well, then they don't exist on the request object, they need to be in the `req.body`

File: scripts/src/models/applications.ts [hunk 4]
```diff
@@ -8,6 +8,7 @@ export type SignInType = "jwt" | "whitelist";
 export type ApplicationConfig = {
 	max_user_wallets: number | null;
 	sign_in_types: SignInType[];
+	blockchain_version: "2" | "3";
```
Comment (nitzantomer) [inline scripts/src/models/applications.ts]: use the `BlockchainVersion` type you defined in the `models/offers` module

**Comment by nitzantomer** [comment]
@yosriz other than the fact that he's trying to access `req.tx_envelope` and `req.network_id` which shouldn't exist, that's not semantic.

**Comment by yosriz** [comment]
@nitzantomer sorry, missed that

File: scripts/src/public/routes/config.ts [hunk 5]
```diff
@@ -4,14 +4,17 @@ import { NextFunction, Request, RequestHandler, Response } from "express";
...
+import { app } from "../app";
 
 const CONFIG = getConfig();
 let JWT_KEYS: KeyMap;
 // one time get config from payment service
 let BLOCKCHAIN: BlockchainConfig;
 
 export async function init() {
-	BLOCKCHAIN = await getBlockchainConfig(getDefaultLogger());
+	BLOCKCHAIN = await getBlockchainConfig("2", getDefaultLogger());
```
Comment (ravitb) [inline scripts/src/public/routes/config.ts]: It is better to parametrize this.

File: scripts/src/public/routes/config.ts [hunk 6]
```diff
@@ -27,11 +30,23 @@ export type ConfigResponse = {
 export const getConfigHandler = async function(req: Request, res: Response, next: NextFunction) {
 	const data: ConfigResponse = {
 		jwt_keys: await getJwtKeys(),
-		blockchain: await getBlockchainConfig(getDefaultLogger()),
+		blockchain: await getBlockchainConfig("2", getDefaultLogger()),
```
Comment (ravitb) [inline scripts/src/public/routes/config.ts]: Same here as well.

File: scripts/src/public/services/payment.ts [hunk 7]
```diff
@@ -117,7 +121,48 @@ export type BlockchainConfig = {
...
-	const res = await client.get(`${config.payment_service}/config`);
+export async function getBlockchainConfig(blockchainVersion: string, logger: LoggerInstance): Promise<BlockchainConfig> {
+	const res = await client.get(`${getPaymentServiceUrl(blockchainVersion)}/config`);
 	return res.data;
 }
+
+export interface WhitelistTransactionRequest {
+	order_id: string;
+	source: string;
+	destination: string;
```
Comment (doodyparizada) [inline scripts/src/public/services/payment.ts]: Maybe keep the API convention of: 
```
	recipient_address: string;
 	sender_address: string;
```

File: scripts/src/public/services/payment.ts [hunk 8]
```diff
@@ -117,7 +121,48 @@ export type BlockchainConfig = {
...
+		destination: order.blockchainData.recipient_address!,
+		amount: order.amount,
+		xdr: tx_envelope,
+		network_id,
+		app_id
+	};
+
+	// whitelist is only for the new payment service
+	const whitelist_response = (await client.post(`${config.new_payment_service}/whitelist`, payload));
+	if (whitelist_response.status === 401) {
```
Comment (doodyparizada) [inline scripts/src/public/services/payment.ts]: I think that when a status code greater than 300 is received by `axios` it raises an error. So this flow will never enter. You need to catch the error and rethrow it if you want to throw a specific error

Comment (Ronserruya) [reply scripts/src/public/services/payment.ts]: Ok, changed it :+1:

File: scripts/src/public/services/orders.ts [hunk 9]
```diff
@@ -78,8 +78,9 @@ export async function getOrder(orderId: string, logger: LoggerInstance): Promise
 	return orderDbToApi(order);
 }
 
-export async function getOpenOrder(orderId: string): Promise<db.MarketplaceOrder| db.ExternalOrder> {
-	const order = await db.Order.getOne(orderId, "opened") as db.MarketplaceOrder | db.ExternalOrder;
+export async function getOrderForWhitelisting(orderId: string): Promise<db.MarketplaceOrder| db.ExternalOrder> {
+	// Get order, regardless of status
```
Comment (yosriz) [inline scripts/src/public/services/orders.ts]: why regardless of status? didn't we agree on open or pending?

Comment (Ronserruya) [reply scripts/src/public/services/orders.ts]: Need to write a new method that accepts a list of statuses, and it really doesn't matter what the status of the order is

**Pull request merged**